src/hotspot/share/opto/predicates.cpp

```diff
@@ -1221,14 +1244,34 @@ class PredicateUsefulMarkerVisitor : public PredicateVisitor {
 
   // We actually mark the associated OpaqueTemplateAssertionPredicate node useful.
   void visit(const TemplateAssertionPredicate& template_assertion_predicate) override {
-    template_assertion_predicate.opaque_node()->mark_useful();
+    if (!_loop_node->is_CountedLoop()) {
+      // Template Assertion Predicate above LoopNode? Must be unrelated because they can only be created above
+      // CountedLoopNodes during Loop Predication or Range Check Elimination.
+      return;
+    }
+
+    mark_template_useful_if_matching_loop(template_assertion_predicate);
+  }
+
+  // If the stored loop node does not match the current loop node from which we iterate from, we found a Template
+  // Assertion Predicate belonging to an already earlier folded loop in the graph. We need to drop this Template
+  // Assertion Predicate because we are no longer splitting a loop which it belongs to. Moreover, if we do not remove
+  // this Template Assertion Predicate, we could wrongly be creating Initialized Assertion Predicates from it at the
+  // new loop which has completely unrelated loop values. These Initialized Assertion Predicates can then fail at
+  // runtime, and we crash by executing a halt instruction.
+  void mark_template_useful_if_matching_loop(const TemplateAssertionPredicate& template_assertion_predicate) const {
+    OpaqueTemplateAssertionPredicateNode* opaque_node = template_assertion_predicate.opaque_node();
+    if (opaque_node->loop_node() == _loop_node) {
+      template_assertion_predicate.opaque_node()->mark_useful();
+    }
   }
```

Comment:
I'm not sure if it makes to split this into two methods, but that's subjective :sweat_smile: 

It seems to me that the code in `visit` is an optimization for what happens in `mark_template_useful_if_matching_loop`, and does not really make sense on its own.

Reply:
The reasons I've split it is the following:
- The bailout for non-counted loops is actually separate to the marking. So I have a two-step algorithm: bailout + marking which can nicely be split.
- Having `mark_template_useful_if_matching_loop()` allows me to quickly read `visit()` and understand what's going on. Additionally, I can put the details about why we do the marking at the method comment for more interested code readers. Without the extracted method, I would probably need to put an extra "mark template useful if matching loop" comment + the 6 lines of comments at `mark_template_useful_if_matching_loop()` into the `visit()` method which makes it harder to grasp.

I would prefer to stick to what I have now - but I admit it's a subjective matter :-)

Reply:
I leave it up to you :)

But could `opaque_node->loop_node() == _loop_node` even be true if we have `!_loop_node->is_CountedLoop()`? or do we actually need a CountedLoop to even have the match?

Reply:
No, it cannot be true. `opaque_node->loop_node()` is a `CountedLoop`. We could have only `opaque_node->loop_node() == _loop_node` in place. I think I first had that as an assertion in place but then turned it into a bail out. But you're right, it would already be covered by the check. Given it's a rare edge case, I guess we can get rid of it. Then the reason above does not apply anymore that we have 2 steps but only 1. Then it does not make sense to split it further. I merged it back together.